// mpdclient.js/lib/Command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Command = (function () {
    /**
     * Creates new command.
     *
     * @param {string} command - Command for execution.
     * @param {*[]} args - Optional arguments for
     *
     * @constructor
     */
    function Command(command, args) {
        if (args === void 0) { args = []; }
        this.command = command;
        this.args = args;
        this.command = command.trim();
    }
    /**
     * Query should have this form:
     *  command\t"arg1"\t"arg2"\t"arg3"\t"argN"\n
     *
     * @return {string} - Built query.
     */
    Command.prototype.buildQuery = function () {
        if (!this.query) {
            this.query = this.args.reduce(function (prev, item) {
                prev.push('"' + item.toString().replace('"', '\'') + '"');
                return prev;
            }, [this.command]).join('\t') + '\n';
        }
        return this.query;
    };
    /**
     * Getter for args prop.
     *
     * @returns {*} - Array of args.
     */
    Command.prototype.getArgs = function () {
        return this.args;
    };
    /**
     * Getter for command prop.
     *
     * @returns {*} - Command string.
     */
    Command.prototype.getCommand = function () {
        return this.command;
    };
    return Command;
}());
exports.MPDCommand = Command;
